thirdparty/hnswlib/hnswlib/hnswalg.h

```diff
@@ -1421,6 +1421,132 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
+                q.push(start_point);
+                used[start_point] = true;
+                if (level > 0)
+                    touched.push_back(start_point);
+                while (!q.empty()) {
+                    tableint j = q.front();
+                    q.pop();
+                    unsigned int* data;
+                    std::unique_lock<std::mutex> lock(link_list_locks_[j]);
+                    data = (unsigned int*)get_linklist_at_level(j, level);
+                    size_t size = getListCount((linklistsizeint*)data);
+                    tableint* datal = (tableint*)(data + 1);
+                    for (size_t k = 0; k < size; k++) {
+                        tableint cand = datal[k];
+                        if (!used[cand]) {
+                            used[cand] = true;
+                            q.push(cand);
+                            if (level > 0)
+                                touched.push_back(cand);
+                        }
+                    }
+                }
+            }
+            start_points = touched;
+
+            for (tableint i = 0; i < cur_element_count; ++i) {
+                if (element_levels_[i] >= level) {
+                    if (!used[i]) {
+                        if (level > 0) {  // for upper level, directly add edges since nodes num is usually small and fast to search its neighbors
+                            repairGraphConnectivity(i, level);
+                        } else {  // for base level, collect the unreachable nodes and repair them concurrently                 
+                            unreached.push_back(i);
+                        }
+                    }
+                }
+            }
+        }
+        return unreached;
+    }
+
+    // add some edges for those unreachable vectors to improve graph connectivity
+    // only call this method after index building
+    void
+    repairGraphConnectivity(tableint cur_c, int level = 0) {
+        size_t m_max = level ? maxM_ : maxM0_;
+        tableint currObj = enterpoint_node_;
+
+        dist_t curdist = calcDistance(cur_c, currObj);
+
+        for (int level_above = maxlevel_; level_above > level; level_above--) {
+            bool changed = true;
+            while (changed) {
+                changed = false;
+                unsigned int* data;
+                std::unique_lock<std::mutex> lock(link_list_locks_[currObj]);
+                data = (unsigned int*)get_linklist(currObj, level_above);
+                int size = getListCount(data);
+                tableint* datal = (tableint*)(data + 1);
+#if defined(USE_PREFETCH)
+                for (int i = 0; i < size; ++i) {
+                    _mm_prefetch(getDataByInternalId(datal[i]), _MM_HINT_T0);
+                }
+#endif
+                for (int i = 0; i < size; i++) {
+                    tableint cand = datal[i];
+                    if (cand < 0 || cand > max_elements_)
+                        throw std::runtime_error("cand error");
```

Comment:
how exactly is this exception handled in the parent code? A client gets 'Unknown error during building an index'?

Reply:
This code is duplicate with hnswlib search upper layer related code, from my understanding, this should never happen. However, there are bunch of runtime error in the hnswalg.h and caller never handle these situations. I will add exception handler in the Add() function